Cauldron/Assets/Scripts/UI/FadeController.cs

```diff
@@ -0,0 +1,73 @@
+﻿using Cysharp.Threading.Tasks;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace CauldronCodebase
+{
+    public class FadeController : MonoBehaviour
+    {
+        [SerializeField] private Image fadeImage;
+        [SerializeField] [Range(0f, 1f)] private float fadeAlpha;
+        [SerializeField] [Tooltip("Fade in seconds")] private float fadeDuration;
+
+        public async UniTask FadeIn(float duration, float targetAlpha)
+        {
+            fadeImage.gameObject.SetActive(true);
+            float elapsedTime = 0;
+
+            while (elapsedTime < duration)
+            {
+                elapsedTime += Time.unscaledDeltaTime;
+                float alpha = Mathf.Lerp(0, 1, elapsedTime / duration * targetAlpha);
+                fadeImage.color = new Color(fadeImage.color.r, fadeImage.color.g, fadeImage.color.b, alpha);
+                await UniTask.Yield();
+            }
+
+        }
+        
+        public async UniTask FadeIn()
+        {
+            fadeImage.gameObject.SetActive(true);
+            float elapsedTime = 0;
+            while (elapsedTime < fadeDuration)
+            {
+                elapsedTime += Time.unscaledDeltaTime;
+                float alpha = Mathf.Lerp(0, 1, elapsedTime / fadeDuration * fadeAlpha);
+                fadeImage.color = new Color(fadeImage.color.r, fadeImage.color.g, fadeImage.color.b, alpha);
+                await UniTask.Yield();
+            }
+        }
+        
+        public async UniTask FadeOut()
+        {
+            float initialAlpha = fadeImage.color.a;
+            float elapsedTime = 0;
+            while (elapsedTime < fadeDuration)
+            {
+                elapsedTime += Time.unscaledDeltaTime;
```

Comment:
This is literally one function copied 4 times)
Make one universal Fade (from, to, duration) and use FadeIn/FadeOut to call it.

Also, it's a good idea to use default values with the ability to override them, but you can make it in a single method with default parameters: FadeIn(duration = -1, target = -1). If any parameter is < 0, use the default value (because alpha and duration can't be < 0).